Split and refactor TimelockAuthorizer tests

ylv-io opened the pull request:
<!-- If this is deployment-related, please go the the Preview tab and select the appropriate sub-template. -->
<!-- Otherwise, delete everything before #Description -->

# Description

This PR splits TimelockAuthorizer tests into the following files for the sake of clarity:
* TimelockAuthorizerActors
* TimelockAuthorizerExecute
* TimelockAuthorizerPermissions
* TimelockAuthorizerRoot

Tests are refactored for better readability at code and output level.
![image](https://user-images.githubusercontent.com/3285000/226506928-653ebca9-e94e-4893-bbdf-a456d2adf933.png)
Two new tests added: `scheduleGrantPermission` and `scheduleRevokePermission`.

<!-- Describe the changes introduced in this pull request. -->
<!-- Include any context necessary for understanding the PR's purpose. -->

## Type of change

- [ ] Bug fix <!-- (non-breaking change which fixes an issue) -->
- [ ] New feature <!-- (non-breaking change which adds functionality) -->
- [ ] Breaking change <!-- (would cause existing functionality to not work as expected) -->
- [ ] Dependency changes
- [x] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests are included for all code paths
- [x] The base branch is either `master`, or there's a description of how to merge

## Issue Resolution

<!-- If this PR addresses an issue, note that here: e.g., Closes/Fixes/Resolves #1346. -->

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -0,0 +1,431 @@
+        const id = await schedule([executor, account]);
+        await advanceTime(delay);
+
+        expect(await authorizer.isExecutor(id, executor)).to.be.true;
+        expect(await authorizer.isExecutor(id, account)).to.be.true;
+
+        const receipt = await authorizer.execute(id, { from: account });
+        expectEvent.inIndirectReceipt(
+          await receipt.wait(),
+          authenticatedContract.interface,
+          'ProtectedFunctionCalled',
+          {
+            data: functionData,
+          }
+        );
```
If you really want to, you can use snapshots to undo the execute and then re-do it from the other account. But tbh I think that's overkill.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -0,0 +1,431 @@
+    });
+  });
+
+  describe('execute', () => {
```
If you're feeling spicy, we could test a scheduled execution where the scheduled action attempts to execute another (or even the same) scheduled action. This will fail because of the reentrancy guard.

ylv-io replied:
This would require another contract which calls authorizer, correct?

nventuro replied:
Yes, it would. But now I think maybe we should remove the guard.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -0,0 +1,431 @@
+    });
+  });
+
+  describe('execute', () => {
```
We should also look at `execute`'s return value, which will be whatever was returned by the target contract. You can query this by doing `authorizer.callStatic.execute(...)`

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -0,0 +1,431 @@
+      return id;
+    };
+
+    it('cancel the action', async () => {
+      const id = await schedule();
+      await authorizer.cancel(id, { from: canceler });
+
+      const scheduledExecution = await authorizer.getScheduledExecution(id);
+      expect(scheduledExecution.cancelled).to.be.true;
+    });
```
I'd do this with both a specific and a global canceler, since we otherwise never have a global canceler use its permission (we only check that `isCanceler` returns true).  We could also add a third test with `root`, which will be able to do it even if it never received the permission.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerRoot.test.ts:
```diff
@@ -0,0 +1,149 @@
+            expect(await authorizer.isPendingRoot(newPendingRoot)).to.be.true;
+          });
+
+          it('emits an event', async () => {
```
I thought this test would check that `RootChangeScheduled` is correctly emitted.

The contents of this test are correct, but they'd fit better under e.g. 'sets the pending root when executed'

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerRoot.test.ts:
```diff
@@ -0,0 +1,149 @@
+        await authorizer.execute(id);
+      });
+
+      context('when the sender is the pending root', async () => {
```
I'd remove this `context`, it adds unnecessary nesting

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerRoot.test.ts:
```diff
@@ -0,0 +1,149 @@
+          await authorizer.claimRoot({ from: user });
+          expect(await authorizer.isPendingRoot(root)).to.be.false;
+          expect(await authorizer.isPendingRoot(user)).to.be.false;
+          expect(await authorizer.isPendingRoot(ZERO_ADDRESS)).to.be.true;
```
We also have `getPendingRoot` and `getRoot` :sweat_smile:

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+        );
+      });
+
+      it('can schedule a grant permission', async () => {
+        const id = await authorizer.scheduleGrantPermission(ACTION_1, user, WHERE_1, [], { from: root });
+
+        // should not be able to execute before delay
+        await expect(authorizer.execute(id, { from: root })).to.be.revertedWith('ACTION_NOT_YET_EXECUTABLE');
+
+        await advanceTime(delay);
+        await authorizer.execute(id, { from: root });
+
+        expect(await authorizer.canPerform(ACTION_1, user, WHERE_1)).to.be.true;
+        expect(await authorizer.canPerform(ACTION_2, user, WHERE_2)).to.be.false;
+      });
```
This is a `scheduleGrantPermission` test, i'd not have it here

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+    });
+
+    context('when there is a no delay set to grant permissions', () => {
+      it('reverts if the sender is not the granter', async () => {
+        await expect(authorizer.grantPermission(ACTION_1, user, WHERE_1, { from: granter })).to.be.revertedWith(
+          'SENDER_IS_NOT_GRANTER'
+        );
+      });
```
This is a bit odd because you're actually using `granter` here, but it's not yet been made a granter. I'd remove this test and check granter status in other places

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+        itGrantsPermissionCorrectly(() => root);
+      });
+
+      context('when the sender is granter', () => {
+        sharedBeforeEach('makes a granter', async () => {
+          await authorizer.addGranter(ACTION_1, granter, EVERYWHERE, { from: root });
+          await authorizer.addGranter(ACTION_2, granter, EVERYWHERE, { from: root });
+        });
+        itGrantsPermissionCorrectly(() => granter);
```
We could add two `it`s here where `granter` is unable to grant permission over a third action. 

It would also be good to have tests for a non global granter, who is unable to grant permissions over _any_ address. We could maybe make `granter` a granter over `WHERE_1` and `WHERE_2`, so that we don't need to change `itGrantsPermissionCorrectly`, and then simply add a few `it`s.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+        const id = await authorizer.scheduleGrantPermission(ACTION_1, user, WHERE_1, [], { from: getSender() });
+
+        await advanceTime(delay);
+        await authorizer.execute(id);
```
You could also test the `PermissionGranted` event here, for completeness

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+      itScheduleGrantPermissionCorrectly(() => root);
+    });
+
+    context('when the sender is granter', () => {
+      sharedBeforeEach('makes a granter', async () => {
+        await authorizer.addGranter(ACTION_1, granter, EVERYWHERE, { from: root });
+        await authorizer.addGranter(ACTION_2, granter, EVERYWHERE, { from: root });
+      });
+      itScheduleGrantPermissionCorrectly(() => granter);
```
Similar to my `grantPermission` suggestion, I'd try to improve handling of scenarios where the action fails because the caller does not have the right permissions.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+        const id = await authorizer.scheduleGrantPermission(ACTION_1, user, WHERE_1, [], { from: getSender() });
+
+        await advanceTime(delay);
+        await authorizer.execute(id);
```
I really like these simple tests for the `cancel` and `execute` flow, without going crazy (e.g. not testing what happens if you execute a cancelled action, since we already test that in `execute` tests), but showing a reasonable happy path for all features.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+        );
+      });
+
+      it('can schedule a revoke permission', async () => {
+        const id = await authorizer.scheduleRevokePermission(ACTION_1, user, WHERE_1, [], { from: root });
+
+        // should not be able to execute before delay
+        await expect(authorizer.execute(id, { from: root })).to.be.revertedWith('ACTION_NOT_YET_EXECUTABLE');
+
+        await advanceTime(delay);
+        await authorizer.execute(id, { from: root });
+
+        expect(await authorizer.canPerform(ACTION_1, user, WHERE_1)).to.be.false;
+        expect(await authorizer.canPerform(ACTION_2, user, WHERE_2)).to.be.false;
+      });
```
Similar to `grant`, this is a `scheduleRevoke` test and I'd remove it.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+
+      function itRevokesPermissionCorrectly(getSender: () => SignerWithAddress) {
+        context('when the user does not have the permission', () => {
+          it('ignores the request and cannot perform the requested action everywhere', async () => {
+            await expect(authorizer.revokePermission(ACTION_1, user, WHERE_1, { from: getSender() })).not.to.be
+              .reverted;
+
```
How do you feel about this? Should we revert instead?

nventuro replied:
I don't think it adds any value, can make things confusing, enables silent user error, and makes tests more complex.

nventuro replied:
Created https://github.com/balancer/balancer-v2-monorepo/issues/2379

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+        });
+      }
+
+      context('when the sender is root', () => {
+        itRevokesPermissionCorrectly(() => root);
+      });
+
+      context('when the sender is revoker', () => {
+        sharedBeforeEach('makes a revoker', async () => {
+          await authorizer.addRevoker(revoker, EVERYWHERE, { from: root });
+        });
+        itRevokesPermissionCorrectly(() => revoker);
+      });
+    });
```
Same suggestions regarding permission checks for this and scheduleRevoke as I submitted for grant and scheduleGrant (since they're all similar)

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+  describe('renouncePermission', () => {
+    context('when the sender does not have the permission', () => {
+      context('when renouncing the permission for a specific contract', () => {
+        it('ignores the request and still cannot perform the requested action everywhere', async () => {
+          await expect(authorizer.renouncePermission(ACTION_1, WHERE_1, { from: user })).not.to.be.reverted;
+
+          expect(await authorizer.canPerform(ACTION_1, user, EVERYWHERE)).to.be.false;
+        });
```
I'd also consider reverting here

ylv-io replied:
I agree but let's make it a separate PR. This one is big enough.

nventuro replied:
Created https://github.com/balancer/balancer-v2-monorepo/issues/2379

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+      });
+    });
+
+    context('when the user has the permission for a specific contract', () => {
```
I'd add a test where `revoke` has a delay (so we can only do `scheduleRevoke`), but `renounce` still works immediately, with no delay.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerPermissions.test.ts:
```diff
@@ -0,0 +1,802 @@
+        );
+      });
+
+      function itGrantsPermissionCorrectly(getSender: () => SignerWithAddress) {
```
This `getSender` thing is a very nice way to handle the async parameter issue we face elsewhere :ok_hand:

ylv-io commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -0,0 +1,431 @@
+      expect(scheduledExecution.executableAt).to.be.at.almostEqual((await currentTimestamp()).add(delay));
+    });
+
+    it('cannot execute the action immediately', async () => {
+      const id = await schedule(authenticatedContract);
+      await expect(authorizer.execute(id)).to.be.revertedWith('ACTION_NOT_YET_EXECUTABLE');
+    });
+
+    it('can be executed by anyone', async () => {
+      const id = await schedule(authenticatedContract);
+      await advanceTime(delay);
+
+      const receipt = await authorizer.execute(id);
+      expectEvent.inReceipt(await receipt.wait(), 'ExecutionExecuted', { scheduledExecutionId: id });
+
+      const scheduledExecution = await authorizer.getScheduledExecution(id);
+      expect(scheduledExecution.executed).to.be.true;
+
+      expectEvent.inIndirectReceipt(await receipt.wait(), authenticatedContract.interface, 'ProtectedFunctionCalled', {
+        data: functionData,
+      });
+    });
+
+    it('cannot be executed twice', async () => {
+      const id = await schedule(authenticatedContract);
+      await advanceTime(delay);
+
+      await authorizer.execute(id);
+      await expect(authorizer.execute(id)).to.be.revertedWith('ACTION_ALREADY_EXECUTED');
+    });
```
They certainly are but we tend to have overlap in tests suites of close functions, for example, schedule/execute, scheduleGrantPermission/grantPermission. I am neutral on this. If you want to remove this let me know.

nventuro replied:
I'd remove these because a) we have the same exact test for `execute`, so they don't really add anything, and b) reading these caused me some confusion as I started wondering what we should test in each `describe` block, thinking about other missing `execute` tests, going to that section and double-checking coverage, etc., which made it harder for me to be confident about correctness of the test suite.

ylv-io commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -0,0 +1,431 @@
+      }
+    });
+
+    it('cannot execute the action immediately', async () => {
+      const id = await schedule(authenticatedContract, [executor]);
+      await expect(authorizer.execute(id, { from: executor })).to.be.revertedWith('ACTION_NOT_YET_EXECUTABLE');
+    });
+
+    it('can be executed by the executor only', async () => {
+      const id = await schedule(authenticatedContract, [executor]);
+      await advanceTime(delay);
+
+      await expect(authorizer.execute(id, { from: user })).to.be.revertedWith('SENDER_IS_NOT_EXECUTOR');
+
+      const receipt = await authorizer.execute(id, { from: executor });
+      expectEvent.inReceipt(await receipt.wait(), 'ExecutionExecuted', { scheduledExecutionId: id });
+
+      const scheduledExecution = await authorizer.getScheduledExecution(id);
+      expect(scheduledExecution.executed).to.be.true;
+
+      expectEvent.inIndirectReceipt(await receipt.wait(), authenticatedContract.interface, 'ProtectedFunctionCalled', {
+        data: functionData,
+      });
+    });
+
+    it('cannot be executed twice', async () => {
+      const id = await schedule(authenticatedContract, [executor]);
+      await advanceTime(delay);
+
+      await authorizer.execute(id, { from: executor });
+      await expect(authorizer.execute(id, { from: executor })).to.be.revertedWith('ACTION_ALREADY_EXECUTED');
+    });
```
Same as above. Do you want to remove these tests?

ylv-io commented:
Thank you for the comments! I've addressed most comments. Going from commit to commit should be a nice experience. I would implement reverts for `revokePermission/renouncePermission` as a separate PR. I also need to do address extra tests for grantPermission, revokePermission, scheduleGrantPermission, and scheduleRevokePermission.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -253,6 +253,12 @@ describe('TimelockAuthorizer execute', () => {
+
+    it('reverts if data is less than 4 bytes', async () => {
+      await expect(authorizer.schedule(authenticatedContract.address, '0x00', [], { from: user })).to.be.revertedWith(
+        'SENDER_DOES_NOT_HAVE_PERMISSION'
```
Hm not a fan of this error message. Let's change `_decodeSelector` so that it reverts if it cannot get a selector.

nventuro replied:
Created https://github.com/balancer/balancer-v2-monorepo/issues/2380

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -165,6 +165,13 @@ describe('TimelockAuthorizer execute', () => {
+    it('emits ExecutionScheduled event', async () => {
+      const receipt = await authorizer.instance.connect(user).schedule(authenticatedContract.address, data, []);
+
+      // _scheduledExecutions doesn't have a getter so we can't fetch it before the call
+      expectEvent.inReceipt(await receipt.wait(), 'ExecutionScheduled', { scheduledExecutionId: 2 });
```
I don't understand this comment, what do you meant?

Also, let's test the `actionId` param of the event.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -0,0 +1,431 @@
+      expect(scheduledExecution.executableAt).to.be.at.almostEqual((await currentTimestamp()).add(delay));
+    });
+
+    it('emits ExecutorAdded events', async () => {
```
Agreed, even if the helper does it today indirectly, let's have an explicit test for it.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -370,6 +370,17 @@ describe('TimelockAuthorizer execute', () => {
+
+      await advanceTime(delay);
+
+      await expect(authorizer.execute(nextId, { from: executor })).to.be.revertedWith('CANNOT_SCHEDULE_AUTHORIZER_ACTIONS');
```
Note that this revert is not what you're trying to test: you're attempting to schedule an authorizer call directly (which is not allowed). We actually already have a test for this.

But I think we should maybe remove the reentrancy guard check altogether.

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerRoot.test.ts:
```diff
@@ -74,11 +74,9 @@ describe('TimelockAuthorizer root', () => {
-            await advanceTime(ROOT_CHANGE_DELAY);
-            const receipt = await authorizer.execute(id);
-            expectEvent.inReceipt(await receipt.wait(), 'PendingRootSet', { pendingRoot: newPendingRoot.address });
+            expectEvent.inReceipt(await receipt.wait(), 'RootChangeScheduled', { newRoot: newPendingRoot.address });
```
This new test is correct, but let's also keep the old test (i.e. that `PendingRootSet` is emitted correctly when executing). Sorry if my previous message was not clear.

nventuro reviewed (commented):
Thank you for carefully addressing each comment in a separate commit! It made reviewing the changes very simple and pleasant :)

ylv-io commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -312,6 +312,17 @@ describe('TimelockAuthorizer execute', () => {
+      await advanceTime(delay);
+      const ret = await authorizer.instance.connect(executor).callStatic.execute(id);
+
+      // we have to slice first 4 selector bytes from the input data to get the return data
+      expect(ret).to.be.eq(
+        '0x' + authenticatedContract.interface.encodeFunctionData('protectedFunction', [functionData]).slice(10)
```
I don't think it is the case. ABI encoding bytes seems to add zeros padding.

nventuro replied:
Well, yes, the abi-encoded bytes :stuck_out_tongue:. I think you can do `defaultAbiCoder.encode([bytes], [functionData])`

nventuro commented on pkg/vault/test/authorizer/TimelockAuthorizerExecute.test.ts:
```diff
@@ -254,6 +254,12 @@ describe('TimelockAuthorizer execute', () => {
+    it('reverts if schedule for EOA', async () => {
+      // we do not specify reason here because call to an EOA results in the following error:
+      // Transaction reverted without a reason
+      await expect(authorizer.schedule(other.address, '0x00', [], { from: user })).to.be.reverted;
```
Related to https://github.com/balancer/balancer-v2-monorepo/issues/2380

nventuro reviewed (approved):
The last round of changes is brilliant :) Thanks a lot for your work!
